src/envoy/auth/jwt.cc

```diff
@@ -0,0 +1,232 @@
+}
+
+const uint8_t *CastToUChar(const std::string &str) {
+  return reinterpret_cast<const uint8_t *>(str.c_str());
+}
+
+bssl::UniquePtr<EVP_PKEY> EvpPkeyFromStr(const std::string &pkey_pem) {
+  std::string pkey_der = Base64::decode(pkey_pem);
+  bssl::UniquePtr<RSA> rsa(
+      RSA_public_key_from_bytes(CastToUChar(pkey_der), pkey_der.length()));
+  if (rsa == nullptr) {
+    return nullptr;
+  }
+  bssl::UniquePtr<EVP_PKEY> key(EVP_PKEY_new());
+  EVP_PKEY_set1_RSA(key.get(), rsa.get());
+
+  return key;
+}
+
+const EVP_MD *EvpMdFromAlg(const std::string &alg) {
+  /*
+   * may use
+   * EVP_sha384() if alg == "RS384" and
+   * EVP_sha512() if alg == "RS512"
+   */
+  if (alg == "RS256") {
+    return EVP_sha256();
+  } else {
+    return nullptr;
+  }
+}
+
+bool VerifySignature(bssl::UniquePtr<EVP_PKEY> key, const std::string &alg,
+                     const uint8_t *signature, size_t signature_len,
+                     const uint8_t *signed_data, size_t signed_data_len) {
+  bssl::UniquePtr<EVP_MD_CTX> md_ctx(EVP_MD_CTX_create());
+  const EVP_MD *md = EvpMdFromAlg(alg);
+
+  if (md == nullptr) {
+    return false;
+  }
+  if (md_ctx == nullptr) {
+    return false;
+  }
+  if (EVP_DigestVerifyInit(md_ctx.get(), nullptr, md, nullptr, key.get()) !=
+      1) {
+    return false;
+  }
+  if (EVP_DigestVerifyUpdate(md_ctx.get(), signed_data, signed_data_len) != 1) {
+    return false;
+  }
+  if (EVP_DigestVerifyFinal(md_ctx.get(), signature, signature_len) != 1) {
+    return false;
+  }
+  return true;
+}
+
+bool VerifySignature(const std::string &pkey_pem, const std::string &alg,
+                     const std::string &signature,
+                     const std::string &signed_data) {
+  return VerifySignature(EvpPkeyFromStr(pkey_pem), alg, CastToUChar(signature),
+                         signature.length(), CastToUChar(signed_data),
+                         signed_data.length());
+}
+
+}  // namespace
+
+namespace Jwt {
+
+std::unique_ptr<rapidjson::Document> Decode(const std::string &jwt,
+                                            const std::string &pkey_pem) {
+  /*
+   * TODO: return failure reason (something like
+   * https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/jwt/jwt_verifier.h#L38)
+   */
+
+  /*
+   * TODO: support jwk
+   */
+
+  /*
+   * jwt must have exactly 2 dots
+   */
+  if (std::count(jwt.begin(), jwt.end(), '.') != 2) {
+    return nullptr;
+  }
+  std::vector<std::string> jwt_split = StringUtil::split(jwt, '.');
+  if (jwt_split.size() != 3) {
+    return nullptr;
+  }
+  const std::string &header_base64url_encoded = jwt_split[0];
+  const std::string &payload_base64url_encoded = jwt_split[1];
+  const std::string &signature_base64url_encoded = jwt_split[2];
+  std::string signed_data = jwt_split[0] + '.' + jwt_split[1];
+
+  /*
+   * verification
+   */
+  rapidjson::Document header_json;
+  if (header_json.Parse(Base64UrlDecode(header_base64url_encoded).c_str())
```

Comment:
header_base64url_encoded is "const string&" type,  but Base64UrlDecode() expecting "string" since the function is modifying it.   Compiler should fail it